tests/request_helpers_test.py

```diff
@@ -0,0 +1,408 @@
+    parent_span_id = '987d98239475e0a8'
+
+    # No B3-headers; return None to get a new local-root span
+    assert None is request_helpers.extract_zipkin_attrs_from_headers(
+        {}, sample_rate=88.2)
+    assert [] == mock_random.mock_calls
+
+    # Just 2 bits is a sample-rate defer; die-roll is "no":
+    mock_random.return_value = 0.883
+    assert ZipkinAttrs(
+        trace_id, span_id, None, '0', False,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+    }, sample_rate=88.2)
+    assert [
+        mock.call()
+    ] == mock_random.mock_calls
+    mock_random.reset_mock()
+
+    # Just 2 bits is a sample-rate defer; die-roll is "yes":
+    mock_random.return_value = 0.881
+    assert ZipkinAttrs(
+        trace_id, span_id, None, '0', True,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+    }, sample_rate=88.2)
+    assert [
+        mock.call()
+    ] == mock_random.mock_calls
+    mock_random.reset_mock()
+
+    # defer with a parent span; die-roll is "no":
+    mock_random.return_value = 0.883
+    assert ZipkinAttrs(
+        trace_id, span_id, parent_span_id, '0', False,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+        'X-B3-ParentSpanId': parent_span_id,
+    }, sample_rate=88.2)
+    assert [
+        mock.call()
+    ] == mock_random.mock_calls
+    mock_random.reset_mock()
+
+    # defer with a parent span; die-roll is "yes":
+    mock_random.return_value = 0.881
+    assert ZipkinAttrs(
+        trace_id, span_id, parent_span_id, '0', True,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+        'X-B3-ParentSpanId': parent_span_id,
+    }, sample_rate=88.2)
+    assert [
+        mock.call()
+    ] == mock_random.mock_calls
+    mock_random.reset_mock()
+
+    # non-defer no-trace, no parent span
+    mock_random.return_value = 0.881  # a "yes" if it were used
+    assert ZipkinAttrs(
+        trace_id, span_id, None, '0', False,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+        'X-B3-Sampled': no_trace_str,
+    }, sample_rate=88.2)
+    assert [] == mock_random.mock_calls
+
+    # non-defer yes-trace, no parent span
+    for yes_trace_str in ('true', '1'):
+        mock_random.return_value = 0.883  # a "no" if it were used
+        assert ZipkinAttrs(
+            trace_id, span_id, None, '0', True,
+        ) == request_helpers.extract_zipkin_attrs_from_headers({
+            'X-B3-TraceId': trace_id,
+            'X-B3-SpanId': span_id,
+            'X-B3-Sampled': yes_trace_str,
+        }, sample_rate=88.2)
+        assert [] == mock_random.mock_calls
+
+    # non-defer debug-trace, no parent span
```

Comment:
there's no `d` in here

Reply:
I went back and double-checked https://github.com/openzipkin/b3-propagation and the "d" thing is only for single-header AFAICT.

For multi-header, it says only (no mention of "d" for Sampling State nor Debug Flag):
```
Sampling State

An accept sampling decision is encoded as X-B3-Sampled: 1 and a deny as X-B3-Sampled: 0. Absent means defer the decision to the receiver of this header. For example, a Sampled header might look like: X-B3-Sampled: 1.

Note: Before this specification was written, some tracers propagated X-B3-Sampled as true or false as opposed to 1 or 0. While you shouldn't encode X-B3-Sampled as true or false, a lenient implementation may accept them.
Debug Flag

Debug is encoded as X-B3-Flags: 1. Debug implies an accept decision, so don't also send the X-B3-Sampled header.
```